lib/spack/spack/directives.py

```diff
@@ -182,10 +192,28 @@ def _depends_on(pkg, spec, when=None):
         when = pkg.name
     when_spec = parse_anonymous_spec(when, pkg.name)
 
+    if deptypes is None:
+        # The default deptype is build and install because the common case is
+        # to build against a library which then turns into a runtime dependency
+        # due to the linker.
+        # XXX(deptype): Add DEP_RUN to this? It's an uncommon dependency type,
+        #               but is most backwards-compatible.
+        deptypes = (DEP_BUILD, DEP_INSTALL)
+
+    if isinstance(deptypes, str):
+        deptypes = (deptypes,)
+
+    for deptype in deptypes:
+        if deptype not in DEP_ALL:
+            raise UnknownDependencyTypeError('depends_on', pkg.name, deptype)
+
     dep_spec = Spec(spec)
     if pkg.name == dep_spec.name:
         raise CircularReferenceError('depends_on', pkg.name)
 
+    pkg_deptypes = pkg._deptypes.setdefault(dep_spec.name, set())
+    map(pkg_deptypes.add, deptypes)
```

Comment:
Also, I've done a decent amount of Haskell, so the higher-order functions are preferred when I write such things. I'll unroll it.